trace/obfuscate.go

```diff
@@ -22,19 +55,148 @@ type obfuscationVisitor struct {
+			key = header.Key
+		case *pb.HTTPMeta_Query:
+			queryParam := httpMeta.GetQuery()
+			key = queryParam.Key
+		}
+		// Check if the key is in the list of keys to obfuscate.
+		if s.obfuscationOptions.SensitiveDataKeys.Contains(strings.ToLower(key)) {
+			return ObfuscatePrimitiveWithRedactedString(d)
+		}
+	}
+
+	primitive := d.GetPrimitive()
+	if primitive == nil {
+		// Not a primitive, regex will be applied to primitive string values only.
+		return Continue
+	}
+
+	stringValue := primitive.GetStringValue()
+	if stringValue == nil {
+		// Not a string, regex will be applied to string values only.
+		return Continue
+	}
+
+	for _, pattern := range s.obfuscationOptions.SensitiveDataValuePatterns {
+		if pattern.MatchString(stringValue.Value) {
+			return ObfuscatePrimitiveWithRedactedString(d)
+		}
+	}
+
+	return Continue
+}
+
+func (s *redactSensitiveInfoVisitor) EnterHTTPBody(self interface{}, ctx vis.SpecVisitorContext, b *pb.HTTPBody) Cont {
+	node, _ := ctx.GetInnermostNode(reflect.TypeOf((*pb.Data)(nil)))
+	data := node.(*pb.Data)
+	if data == nil {
+		return Continue
+	}
+
+	// Traverse the data node and redact sensitive information.
+	s.traverseAndRedactSensitiveInfo(data, false)
+
+	return Continue
+}
+
+// Traverses the protobuf data and redacts sensitive information based on the sensitive keys.
+// The function has 2 parameters:
+//  1. data: The current node in the recursive tree traversal of the protobuf data.
+//  2. redactData: A boolean flag to indicate if the data should be redacted irrespective of it's key and value.
+//     This will be true when the parent node is a sensitive key.
+//
+// The function checks if the data fields are primitive and redacts them if their keys are sensitive.
+// For non-primitive fields, it recursively traverses it's child nodes with the redactData flag set to true if the parent node is sensitive.
+func (s *redactSensitiveInfoVisitor) traverseAndRedactSensitiveInfo(data *pb.Data, redactData bool) {
+	switch val := data.Value.(type) {
+	case *pb.Data_Primitive:
+		if redactData {
+			ObfuscatePrimitiveWithRedactedString(data)
+		}
+	case *pb.Data_Struct:
+		// Traverse the struct's fields and redact sensitive information based on key.
+		structData := val.Struct
+		for fieldName, fieldData := range structData.GetFields() {
+			// Check if the key is sensitive. If it is, redact the child data.
+			redactData := redactData || s.obfuscationOptions.SensitiveDataKeys.Contains(strings.ToLower(fieldName))
+			s.traverseAndRedactSensitiveInfo(fieldData, redactData)
+		}
+	case *pb.Data_List:
+		listData := val.List
+		if items := listData.GetElems(); items != nil {
+			for _, itemData := range items {
+				// Step through each item in the list and traverse the data node.
+				s.traverseAndRedactSensitiveInfo(itemData, redactData)
+			}
+		}
+	default:
+		// Unknown data type, mark as REDACTED string.
+		printer.Errorf("Unknown data type (OneOf or Optional) found, marking as REDACTED string\n")
```

Comment:
How about using the reflect package to show the type here.  It might be something silly like a non-pointer version of a type mentioned above (other code has cases to deal with that) or otherwise unexpected.

`reflect.TypeOf(data.Value).String()`?